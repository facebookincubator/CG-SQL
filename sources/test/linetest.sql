/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

declare proc printf no check;

-- testing line numbers for simple statement boundaries, any statement will do
-- something with no "interior" works best.
create proc based_statements()
begin
  declare x integer not null;
  set x := 1;
  set x := 2;
  set x := 3;
  @echo c, "/* hello "; -- no line info for these!
  @echo c, "world \n";  -- no line info for these!
  set x := 4;
  set x := 5;
end;

-- testing simple if statements in various forms
create proc if_test()
begin
  declare x integer not null;
  if x = 1 then
     set x := 10;
  end if;

  if x = 2 then
     set x := 21;
  else
     set x := 22;
  end if;

  if x = 3 then
     set x := 31;
  else if x = 3  then
     set x := 32;
  else
     set x := 32;
  end if;
end;

-- testing if and case together with a coalesce thrown in
create proc case_if()
begin
  if 1 then
    call printf("one");
  else
    call printf("two");
  end if;

  declare x integer;

  set x :=  coalesce( case 
             when  1 
           then 200
         when 2
   then 300
    end, 
    3000);
end;

-- testing the in predicate by itself
create proc in_pred_lines(i integer not null, out b bool not null )
begin
  set b := i in (
           1,
           3,
           7);
end;

-- testing in predicate mixed with case
create proc in_pred_and_case(i integer not null, out b bool not null )
begin
  set b := case when
    i > 8
    then
      i in (
      10,
      12,
      14)
    else
      i in (
      1,
      3,
      7)
    end;
end;
