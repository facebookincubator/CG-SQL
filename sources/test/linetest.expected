/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

LINE TEST EXPECTED OUTPUT

line numbers are relative to the line the #define _PROC_
anything outside the #define and #undef for _PROC is ignored

The test cases and remember this outside stuff is ignored in this file,
it's just here for your viewing pleasure.  The source of truth is in "test/linetest.sql"
annotated source code from there included.

------------------------ ------------------------ ------------------------

TEST: simple statements
----------------
10: CREATE PROC based_statements ()
11: BEGIN
12:   DECLARE x INTEGER NOT NULL;      THIS IS FOR DISPLAY ONLY.
13:   SET x := 1;                      THIS TEXT IS PROOF THAT THIS IS NOT PARSED.
14:   SET x := 2;
15:   SET x := 3;
16:   @ECHO c, "/* hello ";
17:   @ECHO c, "world \n";
18:   SET x := 4;
19:   SET x := 5;
20: END;
----------------

Note that the proc started at line 10.  That's because there was a comment
but the starting line is not relevant, everything will be normalized to
a proc that starts at line 1 anyway.

#define _PROC_ "based_statements"
# 1
void based_statements(void) {
# 1
  cql_int32 x = 0;
# 1

# 3 "test/linetest.sql"
# 4 "test/linetest.sql"
  x = 1;
# 5 "test/linetest.sql"
  x = 2;
# 6 "test/linetest.sql"
  x = 3;
# 6
  /* hello world 
# 9 "test/linetest.sql"
  x = 4;
# 10 "test/linetest.sql"
  x = 5;
# 11 "test/linetest.sql"

# 11
}
#undef _PROC_
------------------------ ------------------------ ------------------------

TEST: assorted if combinations

------------------------
 1: create proc if_test()
 2: begin
 3:   declare x integer not null;          THIS IS FOR DISPLAY ONLY.
 4:   if x = 1 then                        THIS TEXT IS PROOF THAT THIS IS NOT PARSED.
 5:      set x := 10;
 6:   end if;
 7:
 8:   if x = 2 then
 9:      set x := 21;
10:   else
11:      set x := 22;
12:   end if;
13:
14:   if x = 3 then
15:      set x := 31;
16:   else if x = 3  then
17:      set x := 32;
18:   else
19:      set x := 32;
20:   end if;
21: end;
------------------------

#define _PROC_ "if_test"
# 1
void if_test(void) {
# 1
  cql_int32 x = 0;
# 1

# 3 "x"
# 4 "x"
# 4 "x"
  if (x == 1) {
# 5 "x"
    x = 10;
# 5 "x"
  }
# 8 "x"
# 8 "x"
  if (x == 2) {
# 9 "x"
    x = 21;
# 9 "x"
  }
# 11 "x"
  else {
# 11 "x"
    x = 22;
# 11
  }
# 14 "x"
# 14 "x"
  if (x == 3) {
# 15 "x"
    x = 31;
# 15 "x"
  }
# 15
  else {
# 16 "x"
    if (x == 3) {
# 17 "x"
      x = 32;
# 17 "x"
    }
# 19 "x"
    else {
# 19 "x"
      x = 32;
# 19
    }
# 19
  }
# 21 "x"

# 21
}
#undef _PROC_

------------------------ ------------------------ ------------------------
TEST: some case, coalesce and if statements

----------------
 1: create proc case_if()
 2: begin
 3:   if 1 then
 4:    call printf("one");
 5:  else                          THIS IS FOR DISPLAY ONLY.
 6:    call printf("two");         THIS TEXT IS PROOF THAT THIS IS NOT PARSED.
 7:   end if;
 8:
 9:   declare x integer;
10:
11:   set x :=  coalesce( case
12:              when  1
13:              then 200
14:              when 2
15:              then 300
16:              end,
17:             3000);
18:
19: end;
----------------

#define _PROC_ "case_if"
# 1
void case_if(void) {
# 1
  cql_nullable_int32 x;
# 1
  cql_set_null(x);
# 1
  cql_int32 _tmp_int_0 = 0;
# 1
  cql_nullable_int32 _tmp_n_int_1;
# 1
  cql_set_null(_tmp_n_int_1);
# 1

# 3 "test/linetest.sql"
# 3 "test/linetest.sql"
  if (1) {
# 4 "test/linetest.sql"
    printf("one");
# 4 "test/linetest.sql"
  }
# 6 "test/linetest.sql"
  else {
# 6 "test/linetest.sql"
    printf("two");
# 6
  }
# 9 "test/linetest.sql"
# 11 "test/linetest.sql"
# 11 "test/linetest.sql"
  do {
# 16 "test/linetest.sql"
# 12 "test/linetest.sql"
    do {
# 12 "test/linetest.sql"
      if (1) {
# 13 "test/linetest.sql"
        cql_set_notnull(_tmp_n_int_1, 200);
# 13
        break;
# 13
      }
# 14 "test/linetest.sql"
      if (2) {
# 15 "test/linetest.sql"
        cql_set_notnull(_tmp_n_int_1, 300);
# 15
        break;
# 15
      }
# 16 "test/linetest.sql"
      cql_set_null(_tmp_n_int_1);
# 16
    } while (0);
# 16
    if (!_tmp_n_int_1.is_null) {
# 16
      _tmp_int_0 = _tmp_n_int_1.value;
# 16
      break;
# 16
    }
# 17 "test/linetest.sql"
    _tmp_int_0 = 3000;
# 17
  } while (0);
# 17
  cql_set_notnull(x, _tmp_int_0);
# 18 "test/linetest.sql"

# 18
}
#undef _PROC_

------------------------ ------------------------ ------------------------


TEST: IN predicate
----------------
1: create proc in_pred_lines(i integer not null, out b bool not null )
2: begin
3:   set b := i in (
4:            1,
5:            3,
6:            7);
7: end;
----------------

#define _PROC_ "in_pred_lines"
# 1
void in_pred_lines(cql_int32 i, cql_bool *_Nonnull b) {
# 1
  cql_contract_argument_notnull((void *)b, 2);
# 1

# 1
  cql_bool _tmp_bool_0 = 0;
# 1
  cql_int32 _tmp_int_2 = 0;
# 1

# 1
  *b = 0; // set out arg to non-garbage
# 3 "x"
  do {
# 3 "x"
    _tmp_int_2 = i;
# 3
    _tmp_bool_0 = 1;
# 4 "x"
    if (_tmp_int_2 == 1) break;
# 5 "x"
    if (_tmp_int_2 == 3) break;
# 6 "x"
    if (_tmp_int_2 == 7) break;
# 6
    _tmp_bool_0 = 0;
# 6 "x"
  } while (0);
# 6
  *b = _tmp_bool_0;
# 7 "x"

# 7
}
#undef _PROC_

----------------
 1: create proc in_pred_and_case(i integer not null, out b bool not null )
 2: begin
 3:   set b := case when
 4:     i > 8
 5:     then
 6:       i in (
 7:       10,
 8:       12,
 9:       14)
10:     else
11:       i in (
12:       1,
13:       3,
14:       7)
15:     end;
16: end;
----------------

#define _PROC_ "in_pred_and_case"
# 1
void in_pred_and_case(cql_int32 i, cql_bool *_Nonnull b) {
# 1
  cql_contract_argument_notnull((void *)b, 2);
# 1

# 1
  cql_bool _tmp_bool_2 = 0;
# 1
  cql_int32 _tmp_int_4 = 0;
# 1
  cql_bool _tmp_bool_1 = 0;
# 1
  cql_int32 _tmp_int_3 = 0;
# 1

# 1
  *b = 0; // set out arg to non-garbage
# 3 "x"
# 4 "x"
  do {
# 4 "x"
    if (i > 8) {
# 6 "x"
      do {
# 6 "x"
        _tmp_int_4 = i;
# 6
        _tmp_bool_2 = 1;
# 7 "x"
        if (_tmp_int_4 == 10) break;
# 8 "x"
        if (_tmp_int_4 == 12) break;
# 9 "x"
        if (_tmp_int_4 == 14) break;
# 9
        _tmp_bool_2 = 0;
# 9 "x"
      } while (0);
# 9
      *b = _tmp_bool_2;
# 9
      break;
# 9
    }
# 11 "x"
    do {
# 11 "x"
      _tmp_int_3 = i;
# 11
      _tmp_bool_1 = 1;
# 12 "x"
      if (_tmp_int_3 == 1) break;
# 13 "x"
      if (_tmp_int_3 == 3) break;
# 14 "x"
      if (_tmp_int_3 == 7) break;
# 14
      _tmp_bool_1 = 0;
# 14 "x"
    } while (0);
# 15 "x"
    *b = _tmp_bool_1;
# 15
  } while (0);
# 16 "x"

# 16
}
#undef _PROC_
